### PR TITLE
refactor: config option parent -> parents

### LINKED ISSUE #26599: Refactor: config option parent -> parents
### Describe the proposed change(s).

Change this to be an array of strings: https://github.com/renovatebot/renovate/blob/8cf2d21082a8fe40590ec3f48b073891cb1c1a30/lib/config/types.ts#L397-L402

+ associated downstream code changes

### PR DESCRIPTION
**Pull request by RahulGautamSingh**
## Changes
Change parent option to be an array of strings, and the associated changes

## Context
Closes: https://github.com/renovatebot/renovate/issues/26599

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/config/types.ts [hunk 1]
```diff
@@ -394,13 +394,7 @@ export interface RenovateOptionBase {
...
   name: string;
 
-  parent?:
-    | 'customDatasources'
-    | 'hostRules'
-    | 'packageRules'
-    | 'postUpgradeTasks'
-    | 'customManagers';
-
+  parents?: string[];
```
Comment (viceice) [inline lib/config/types.ts:397]: should we create a more narrow type? 🤔

File: lib/config/validation.ts [hunk 2]
```diff
@@ -29,7 +29,7 @@ import * as managerValidator from './validation-helpers/managers';
 const options = getOptions();
 
 let optionTypes: Record<string, RenovateOptions['type']>;
-let optionParents: Record<string, RenovateOptions['parent']>;
+let optionParents: Record<string, RenovateOptions['parents']>;
```
Comment (viceice) [inline lib/config/validation.ts:32]:
```suggestion
let optionParents: Record<string, string[]>;
```
it's now a string array

File: lib/config/validation.ts [hunk 3]
```diff
@@ -199,10 +199,12 @@ export async function validateConfig(
       if (
         !isPreset &&
         optionParents[key] &&
-        optionParents[key] !== parentName
+        !optionParents[key]?.includes(parentName)
```
Comment (viceice) [inline lib/config/validation.ts:202]:
```suggestion
        !optionParents[key].includes(parentName)
```
validated right before

File: lib/config/types.ts (hunk 1 above)
Comment (RahulGautamSingh) [reply lib/config/types.ts:397]: I started with this but changed it after seeing the issue description. I think its better to use a narrower type, makes catching issues easier.
 
 ```ts
type AllowedParents = | 'customManagers' | 'customDatasources' | 'hostRules' | 'postUpgradeTasks' | 'packageRules';

parents?: AllowedParents[];
 ```

Comment (rarkins) [reply lib/config/types.ts:397]: Yes

File: lib/config/validation.ts [hunk 4]
```diff
@@ -199,10 +199,12 @@ export async function validateConfig(
       if (
         !isPreset &&
         optionParents[key] &&
-        optionParents[key] !== parentName
+        !optionParents[key].includes(parentName)
       ) {
         // TODO: types (#22198)
-        const message = `${key} should only be configured within a "${optionParents[key]}" object. Was found in ${parentName}`;
+        const message = `${key} should only be configured within a "${optionParents[
```
Comment (rarkins) [inline lib/config/validation.ts:205]:
```suggestion
        const message = `${key} should only be configured within one of "${optionParents[
```

File: lib/config/validation.ts [hunk 5]
```diff
@@ -199,10 +199,12 @@ export async function validateConfig(
...
+          key
+        ]?.join(',')}" object. Was found in ${parentName}`;
```
Comment (rarkins) [inline lib/config/validation.ts:207]:
```suggestion
        ]?.join(',')}" objects. Was found in ${parentName}`;
```

**Review by rarkins — CHANGES_REQUESTED** [review]
Please manually verify the diff in JSON schema and report it here

**Comment by RahulGautamSingh** [comment]
> Please manually verify the diff in JSON schema and report it here

There will no difference, since the parents haven't been modified.
Verified manually. I have posted the json schema generated from this branch [here](https://pastebin.com/ekddmAeK).

File: lib/config/validation.ts [hunk 6]
```diff
@@ -224,10 +225,12 @@ export async function validateConfig(
...
         !isPreset &&
         optionParents[key] &&
-        optionParents[key] !== parentName
+        !optionParents[key]?.includes(parentName as AllowedParents)
       ) {
         // TODO: types (#22198)
-        const message = `${key} should only be configured within a "${optionParents[key]}" object. Was found in ${parentName}`;
+        const message = `${key} should only be configured within one of "${optionParents[
+          key
+        ]?.join(',')}" objects. Was found in ${parentName}`;
```
Comment (rarkins) [inline lib/config/validation.ts:233]:
```suggestion
        ]?.join(' or ')}" objects. Was found in ${parentName}`;
```

File: lib/config/types.ts [hunk 7]
```diff
@@ -368,6 +368,12 @@ export interface ValidationMessage {
   message: string;
 }
 
+export type AllowedParents =
```
Comment (viceice) [inline lib/config/types.ts:371]: we can later add a `$root` item to indicate it's a global option. then we can later set parents for `password` to `$root,hostRule` for better validation and schema generation.

File: lib/config/validation.ts (hunk 2 above)
Comment (viceice) [reply lib/config/validation.ts:36]: ☝️ Updated

**Review by viceice — COMMENTED** [review]
otherwise LGTM

**Review by rarkins — DISMISSED** [review]

**Review by viceice — APPROVED** [review]

**Pull request merged**